README.md
# attestation-honneur-ue-prive
ENGAGEMENT SUR L’HONNEUR A SE SOUMETTRE AUX REGLES RELATIVESA L’ENTREE SUR LE TERRITOIRE NATIONAL METROPOLITAIN DEPUIS UN PAYS DE L’ESPACE EUROPEEN


CHANGELOG.md
# Changelog

## [1.0.0]

- Création Engagement sur l’honneur déplacement depuis l'espace européen


src/js/form.js
import 'bootstrap/dist/css/bootstrap.min.css'

import '../css/main.css'

import formData from '../form-data.json'

import { $, appendTo, createElement, simpleCreateElement } from './dom-utils'

const createTitle = () => {
  const h2 = createElement('h2', { className: 'titre-2', innerHTML: 'Remplissez en ligne votre déclaration numérique : ' })
  const p = createElement('p', { className: 'msg-info', innerHTML: 'Tous les champs sont obligatoires.' })
  return [h2, p]
}

const createFormGroup = ({
  autocomplete = false,
  autofocus = false,
  inputmode,
  label,
  max,
  min,
  maxlength,
  minlength,
  name,
  pattern,
  placeholder = '',
  type = 'text',
}) => {
  const formGroup = createElement('div', { className: 'form-group' })
  const labelAttrs = {
    for: `field-${name}`,
    id: `field-${name}-label`,
    innerHTML: label,
  }
  const labelEl = createElement('label', labelAttrs)

  const inputGroup = createElement('div', { className: 'input-group align-items-center' })
  const inputAttrs = {
    autocomplete,
    autofocus,
    className: 'form-control',
    id: `field-${name}`,
    inputmode,
    min,
    max,
    minlength,
    maxlength,
    name,
    pattern,
    placeholder,
    required: true,
    type,
  }

  const input = createElement('input', inputAttrs)

  const validityAttrs = {
    className: 'validity',
  }
  const validity = createElement('span', validityAttrs)

  const example = createElement('p', { className: 'exemple  basis-100' })

  const appendToFormGroup = appendTo(formGroup)
  appendToFormGroup(labelEl)
  appendToFormGroup(inputGroup)

  const appendToInputGroup = appendTo(inputGroup)
  appendToInputGroup(input)
  appendToInputGroup(validity)
  appendToInputGroup(example)

  return formGroup
}

export function createForm () {
  const form = $('#form-profile')
  // Évite de recréer le formulaire s'il est déjà créé par react-snap (ou un autre outil de prerender)
  if (form.innerHTML !== '') {
    return
  }

  const appendToForm = appendTo(form)

  const formFirstPart = formData
    .flat(1)
    .filter(field => !field.isHidden)
    .map((field,
      index) => {
      const formGroup = createFormGroup({
        autofocus: index === 0,
        ...field,
        name: field.key,
      })

      return formGroup
    })

  const symptomList = [
    'de la fièvre ou des frissons,',
    'une toux ou une augmentation de ma toux habituelle,',
    'une fatigue inhabituelle,',
    'un essoufflement inhabituel quand je parle ou je fais un petit e­ffort,',
    'des douleurs musculaires et/ou des courbatures inhabituelles,',
    'des maux de tête inexpliqués,',
    'une perte de goût ou d’odorat,',
    'des diarrhées inhabituelles.',
  ]
  const createUl = simpleCreateElement('ul')
  const createLi = simpleCreateElement('li')
  const engagementSymptomList = symptomList.map((el) => createLi(el))
  const liste = createUl(...engagementSymptomList)

  const engagementPart = () => {
    const engagementSymptoms = createElement('p', { className: 'symptoms', innerHTML: '<strong>déclare sur l’honneur</strong> n’avoir présenté, au cours des dernières 48 heures, aucun des symptômes suivants :' })
    const engagementSymptomsList = liste
    const engagementContact = createElement('p', { className: 'contact', innerHTML: '<strong>déclare sur l’honneur</strong> ne pas avoir connaissance d’avoir été en contact avec un cas confirmé de covid-19 dans les quatorze jours précédant le départ ;' })
    const engagementExamBio = createElement('p', { className: 'exam-bio', innerHTML: '<strong>m’engage sur l’honneur</strong> à me soumettre à un test antigénique ou à un examen biologique éventuel à l’arrivée.' })
    return [engagementSymptoms, engagementSymptomsList, engagementContact, engagementExamBio]
  }

  appendToForm([...createTitle(), ...formFirstPart, ...engagementPart()])
}


src/js/dom-utils.js
export const $ = (...args) => document.querySelector(...args)
export const $$ = (...args) => [...document.querySelectorAll(...args)]

const plainAttributes = [
  'for',
  'inputmode',
  'minlength',
  'maxlength',
  'min',
  'max',
  'pattern',
]

export const createElement = (tag, attrs) => {
  const el = document.createElement(tag)
  plainAttributes.forEach(plainAttr => {
    if (attrs && plainAttr in attrs && attrs[plainAttr]) {
      el.setAttribute(plainAttr, attrs[plainAttr])
    }
    if (attrs) {
      delete attrs[plainAttr]
    }
  })
  Object.assign(el, attrs)
  return el
}

export const simpleCreateElement = (tag) => (...nodes) => {
  const el = document.createElement(tag)
  el.append(...nodes)
  return el
}

export const appendTo = el => domNodes => {
  if (domNodes[Symbol.iterator]) {
    el.append(...domNodes)
    return
  }
  el.append(domNodes)
}

export function downloadBlob (blob, fileName) {
  const link = createElement('a')
  const url = URL.createObjectURL(blob)
  link.href = url
  link.download = fileName
  document.body.appendChild(link)
  link.click()
}


src/js/form-util.js
import removeAccents from 'remove-accents'

import { $, $$, downloadBlob } from './dom-utils'
import { addSlash, getFormattedDate } from './util'
import { generatePdf } from './pdf-util'
import SecureLS from 'secure-ls'
const secureLS = new SecureLS({ encodingType: 'aes' })
const clearDataSnackbar = $('#snackbar-cleardata')
const storeDataInput = $('#field-storedata')
const conditions = {
  '#field-firstname': {
    length: 1,
  },
  '#field-lastname': {
    length: 1,
  },
  '#field-birthday': {
    pattern: /^([0][1-9]|[1-2][0-9]|30|31)\/([0][1-9]|10|11|12)\/(19[0-9][0-9]|20[0-1][0-9]|2020)/g,
  },
  '#field-birthplace': {
    length: 1,
  },
  '#field-address': {
    length: 1,
  },
  '#field-city': {
    length: 1,
  },
  '#field-zipcode': {
    pattern: /\d{5}/g,
  },
}

function validateAriaFields () {
  return Object.keys(conditions)
    .map((field) => {
      const fieldData = conditions[field]
      const pattern = fieldData.pattern
      const length = fieldData.length
      const isInvalidPattern = pattern && !$(field).value.match(pattern)
      const isInvalidLength = length && !$(field).value.length

      const isInvalid = !!(isInvalidPattern || isInvalidLength)

      $(field).setAttribute('aria-invalid', isInvalid)
      if (isInvalid) {
        $(field).focus()
      }
      return isInvalid
    })
    .includes(true)
}

function updateSecureLS (formInputs) {
  if (wantDataToBeStored() === true) {
    secureLS.set('profile', getProfile(formInputs))
  } else {
    clearSecureLS()
  }
}

function clearSecureLS () {
  secureLS.clear()
}

function clearForm () {
  const formProfile = $('#form-profile')
  formProfile.reset()
  storeDataInput.checked = false
}

function showSnackbar (snackbarToShow, showDuration = 6000) {
  snackbarToShow.classList.remove('d-none')
  setTimeout(() => snackbarToShow.classList.add('show'), 100)

  setTimeout(function () {
    snackbarToShow.classList.remove('show')
    setTimeout(() => snackbarToShow.classList.add('d-none'), 500)
  }, showDuration)
}

export function wantDataToBeStored () {
  return storeDataInput.checked
}

export function setReleaseDateTime (releaseDateInput) {
  const loadedDate = new Date()
  releaseDateInput.value = getFormattedDate(loadedDate)
}

export function toAscii (string) {
  if (typeof string !== 'string') {
    throw new Error('Need string')
  }
  const accentsRemoved = removeAccents(string)
  const asciiString = accentsRemoved.replace(/[^\x00-\x7F]/g, '') // eslint-disable-line no-control-regex
  return asciiString
}

export function getProfile (formInputs) {
  const fields = {}
  for (const field of formInputs) {
    let value = field.value
    if (field.id === 'field-datesortie') {
      const dateSortie = field.value.split('-')
      value = `${dateSortie[2]}/${dateSortie[1]}/${dateSortie[0]}`
    }

    if (typeof value === 'string') {
      value = toAscii(value)
    }
    fields[field.id.substring('field-'.length)] = value
  }
  return fields
}

export function prepareInputs (formInputs, snackbar) {
  const lsProfile = secureLS.get('profile')

  // Continue to store data if already stored
  storeDataInput.checked = !!lsProfile
  formInputs.forEach((input) => {
    if (input.name && lsProfile) {
      input.value = lsProfile[input.name]
    }
    const exempleElt = input.parentNode.parentNode.querySelector('.exemple')
    if (input.placeholder && exempleElt) {
      input.addEventListener('input', (event) => {
        if (input.value) {
          updateSecureLS(formInputs)
          exempleElt.innerHTML = 'ex.&nbsp;: ' + input.placeholder
        } else {
          exempleElt.innerHTML = ''
        }
      })
    }
  })

  $('#field-birthday').addEventListener('keyup', function (event) {
    event.preventDefault()
    const input = event.target
    const key = event.keyCode || event.charCode
    if (key !== 8 && key !== 46) {
      input.value = addSlash(input.value)
    }
  })

  $('#cleardata').addEventListener('click', () => {
    clearSecureLS()
    clearForm()
    showSnackbar(clearDataSnackbar, 3000)
  })
  $('#field-storedata').addEventListener('click', () => {
    updateSecureLS(formInputs)
  })
  $('#generate-btn').addEventListener('click', async (event) => {
    event.preventDefault()

    const invalid = validateAriaFields()
    if (invalid) {
      return
    }
    updateSecureLS(formInputs)
    const pdfBlob = await generatePdf(getProfile(formInputs))

    const creationInstant = new Date()
    const creationDate = creationInstant.toLocaleDateString('fr-CA')
    const creationHour = creationInstant
      .toLocaleTimeString('fr-FR', { hour: '2-digit', minute: '2-digit' })
      .replace(':', '-')

    downloadBlob(pdfBlob, `attestation-${creationDate}_${creationHour}.pdf`)
    showSnackbar(snackbar, 6000)
  })
}

export function prepareForm () {
  const formInputs = $$('#form-profile input')
  const snackbar = $('#snackbar')
  prepareInputs(formInputs, snackbar)
}


src/js/pdf-util.js
import { generateQR } from './util'
import { PDFDocument, PageSizes, rgb } from 'pdf-lib'
import fontkit from '@pdf-lib/fontkit'
import imageUrlRF from '../RF.png'
import imageUrlTAC from '../TAC.png'
import LucioleFontBase from '../Luciole-Regular.ttf'
import LucioleBoldFontBase from '../Luciole-Bold.ttf'
import LucioleItalicFontBase from '../Luciole-Regular-Italic.ttf'
import LucioleBoldItalicFontBase from '../Luciole-Bold-Italic.ttf'
import engagementPdfData from '../engagement-honneur-ue-pdf-data.json'

// const pdfWidth = 595.28
// const pdfHeight = 841.89
// const milimeterWidth = 210
// const milimeterHeight = 297
// const pixelWidth = 892
// const milimeterRatio = 0.35277516462841
const pixelRatio = 1.49845450880258
const sizeRatio = 0.66
export async function generatePdf (profile) {
  const pdfData = engagementPdfData

  const title = 'COVID-19 - Engagement sur l\'honneur'
  const subtitle = 'Engagement sur l\'honneur à se soumettre aux règles relatives à l\'entrée sur le territoire national métropolitain depuis un pays de l\'espace européen'
  // const qrTitle = 'COVID 19 Engagement sur l honneur'
  // const qrSubtitle = 'a se soumettre aux regles relatives a l entree sur le territoire national metropolitain depuis un pays de l espace europeen'
  const creationInstant = new Date()
  const creationDate = creationInstant.toLocaleDateString('fr-FR')
  const creationHour = creationInstant
    .toLocaleTimeString('fr-FR', { hour: '2-digit', minute: '2-digit' })
    .replace(':', 'h')
  profile.creationDate = creationDate
  profile.creationHour = creationHour

  const {
    lastname,
    firstname,
    birthday,
    birthplace,
    address,
    zipcode,
    city,
    creationLocation,
  } = profile

  const data = [
    `Cree le: ${creationDate} - ${creationHour} a ${creationLocation}`,
    `Nom: ${lastname}`,
    `Prenom: ${firstname}`,
    `Naissance: ${birthday}`,
    `Lieu de naissance: ${birthplace}`,
    `Adresse: ${address} ${zipcode} ${city}`,
    `Sortie: cree le ${creationDate} - ${creationHour} a ${creationLocation}`,
    'Motifs: Voyage UE',
    '', // Pour ajouter un ; aussi au dernier élément
  ].join(';\n')

  const pdfDoc = await PDFDocument.create()
  pdfDoc.registerFontkit(fontkit)
  const page = pdfDoc.addPage(PageSizes.A4)
  const height = page.getSize().height

  // set pdf metadata
  pdfDoc.setTitle(title)
  pdfDoc.setSubject(subtitle)
  pdfDoc.setKeywords([
    'covid19',
    'covid-19',
    'attestation',
    'déclaration',
    'déplacement',
    'espace européen',
    'entrée sur le territoire national métropolitain',
    'officielle',
    'gouvernement',
  ])
  pdfDoc.setProducer('DNUM/SDIT')
  pdfDoc.setCreator('')
  pdfDoc.setAuthor("Ministère de l'intérieur")

  let imageBuffer = await fetch(imageUrlRF).then(res => res.arrayBuffer())
  const imageRF = await pdfDoc.embedPng(imageBuffer)
  imageBuffer = await fetch(imageUrlTAC).then(res => res.arrayBuffer())
  const imageTAC = await pdfDoc.embedPng(imageBuffer)

  let fontBuffer = await fetch(LucioleFontBase).then(res => res.arrayBuffer())
  const fontLuciole = await pdfDoc.embedFont(fontBuffer)
  fontBuffer = await fetch(LucioleBoldFontBase).then(res => res.arrayBuffer())
  const fontLucioleBold = await pdfDoc.embedFont(fontBuffer)
  fontBuffer = await fetch(LucioleItalicFontBase).then(res => res.arrayBuffer())
  const fontLucioleItalic = await pdfDoc.embedFont(fontBuffer)
  fontBuffer = await fetch(LucioleBoldItalicFontBase).then(res => res.arrayBuffer())
  const fontLucioleBoldItalic = await pdfDoc.embedFont(fontBuffer)

  let x = 0
  let size = 0
  let block = 0
  let sameline = false
  let font = fontLuciole

  page.drawImage(imageRF, { x: 30, y: page.getHeight() - 80, width: 56, height: 50 })
  page.drawImage(imageTAC, { x: page.getWidth() - 66, y: page.getHeight() - 80, width: 33, height: 50 })

  let y = height - 75

  pdfData.forEach(item => {
    x = 0
    size = Number((item.size * sizeRatio).toFixed(2)) || 10
    block = (item.block === undefined ? 1 : item.block) // Pas de bloc défini dans le JSON = cofficient interligne 1 = standard
    sameline = (item.sameline === undefined ? false : item.sameline) // Pas de flag défini dans le JSON = on passe à la ligne suivante avant d'écrire (fonctionnement standard)

    switch (item.font) {
      case 'LucioleBold':
        font = fontLucioleBold
        break
      case 'LucioleItalic':
        font = fontLucioleItalic
        break
      case 'LucioleBoldItalic':
        font = fontLucioleBoldItalic
        break
      default:
        font = fontLuciole
    }

    if (item.top) {
      x = item.left / pixelRatio
    }
    const label = item.label || ''
    let text = item.variablesNames ? item.variablesNames.reduce((acc, name) => acc + ' ' + profile[name], label) : label

    if (item.type === 'text') {
      if (!sameline) {
        y = y - (block * size) - 5
      }
      page.drawText(text, { x, y, size, font })
    }
    if (item.type === 'input') {
      if (!sameline) {
        y = y - (block * size) - 5
      }
      text = item.inputs.reduce((acc, cur) => acc + ' ' + profile[cur], text)
      page.drawText(text, { x, y, size, font })
    }
  })

  const qrTitle1 = 'QR-code contenant les informations '
  const qrTitle2 = 'de votre attestation numérique'

  const generatedQR = await generateQR(data)

  const qrImage = await pdfDoc.embedPng(generatedQR)
  const pageX0 = pdfDoc.getPages()[0]
  pageX0.drawText(qrTitle1 + '\n' + qrTitle2, { x: 470, y, size: 6, font, lineHeight: 10, color: rgb(1, 1, 1) })

  pageX0.drawImage(qrImage, {
    x: pageX0.getWidth() - 107,
    y: y - 105,
    width: 82,
    height: 82,
  })

  pdfDoc.addPage()
  const pageX = pdfDoc.getPages()[1]
  pageX.drawText(qrTitle1 + qrTitle2, { x: 50, y: pageX.getHeight() - 70, size: 11, font, color: rgb(1, 1, 1) })
  pageX.drawImage(qrImage, {
    x: 50,
    y: pageX.getHeight() - 390,
    width: 300,
    height: 300,
  })

  const pdfBytes = await pdfDoc.save()

  return new Blob([pdfBytes], { type: 'application/pdf' })
}
